internal/plugintest/terraform_json_buffer.go

```diff
@@ -0,0 +1,125 @@
+package plugintest
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"io"
+	"log"
+	"regexp"
+
+	tfjson "github.com/hashicorp/terraform-json"
+)
+
+type TerraformJSONDiagnostics []tfjson.Diagnostic
+
+func (d TerraformJSONDiagnostics) Contains(r *regexp.Regexp, severity tfjson.DiagnosticSeverity) bool {
+	for _, v := range d {
+		if v.Severity != severity {
+			continue
+		}
+
+		if !r.MatchString(v.Summary) && !r.MatchString(v.Detail) {
+			continue
+		}
+
+		return true
+	}
+
+	return false
+}
+
+var _ io.Writer = &TerraformJSONBuffer{}
+var _ io.Reader = &TerraformJSONBuffer{}
+
+// TerraformJSONBuffer is used for storing and processing streaming
+// JSON output generated when running terraform commands with the
+// `-json` flag.
+type TerraformJSONBuffer struct {
+	buf         *bytes.Buffer
+	diagnostics TerraformJSONDiagnostics
+	jsonOutput  []string
+	rawOutput   string
+	parsed      bool
+}
+
+func NewTerraformJSONBuffer() *TerraformJSONBuffer {
+	return &TerraformJSONBuffer{
+		buf: new(bytes.Buffer),
+	}
+}
+
+func (b *TerraformJSONBuffer) Write(p []byte) (n int, err error) {
+	if b.buf == nil {
+		log.Fatal("call NewTerraformJSONBuffer to initialise buffer")
+	}
+
+	return b.buf.Write(p)
+}
+
+func (b *TerraformJSONBuffer) Read(p []byte) (n int, err error) {
+	if b.buf == nil {
+		log.Fatal("call NewTerraformJSONBuffer to initialise buffer")
+	}
+
+	return b.buf.Read(p)
+}
+
+func (b *TerraformJSONBuffer) Parse() {
+	if b.buf == nil {
+		log.Fatal("call NewTerraformJSONBuffer to initialise buffer")
+	}
+
+	scanner := bufio.NewScanner(b.buf)
```

Comment:
Do we need to worry about the 64KB [`bufio.MaxScanTokenSize`](https://pkg.go.dev/bufio#MaxScanTokenSize)? If Terraform can send lines larger than the default buffer size, we may need to provide our own to match Terraform's largest possible line size.

Reply:
Have switched to using [bufio.NewReader](https://pkg.go.dev/bufio#NewReader).